_posts/2019-04-06-trees-and-rectangles-reconciling-sql-and-recursive-programming-languages.md

```diff
@@ -0,0 +1,159 @@
+---
+layout: post
+title: "trees and rectangles: reconciling SQL and recursive programming languages"
+description: ""
+category:
+tags: [haskell, sql]
+---
+
+There's a contradiction at the heart of web development.
+
+Most structures from HTML through to JSON calls are fundamentally
+hierarchical. Relationships are expressed primarily through
+embedding, and this means we mostly deal with trees.
+
+On the other hand, traditional SQL databases talk about
+rectangles. Data is essentially tabular, and even with complicated
+joins and filters, the output is still a rectangle.
+
+How do we resolve this?
+
+### Object-Relational Mappers (or: teaching trees to think about rectangles)
+
+ORMs were one stab at this problem. The underlying idea was that you
+could teach the object system about the structure of a relational
+database, and then make database queries lazily when the pattern of
+calls required the data, thereby letting the developer stay primarily
+in a hierarchical pattern of thought.
+
+While this does sound appealing, it turns out to have many problems in
+practice, from the
+[n+1 query problem](https://medium.com/@bretdoucette/n-1-queries-and-how-to-avoid-them-a12f02345be5)
+to nasty interactions between data cached in objects and the underlying database state. I won't go
+into this too deeply, as [Ted Neward](http://blogs.tedneward.com/post/the-vietnam-of-computer-science/)
+and [Mark Rickerby](https://maetl.net/talks/rise-and-fall-of-orm) have
+already covered it admirably, but in essence: it isn't feasible to
+ignore the structures and guarantees of the underlying relational
+database if you want either correctness or performance. Rails still
+uses this pattern, but even there the Arel package is used to allow
+devs to talk about queries in a way that maps closely to the
+database.
+
+
+### NoSQL! (kill all the rectangles, talk trees)
+
+To massively oversimplify, the term "NoSQL" became popular around 2009
+to describe a disparate group of databases that did not rely primarily
+on the relational view of data as rectangles. Arguably, the earlier object
```

Comment:
Stonebraker and Hellerstein refer to the "Hierarchical Model" as the first attempt at a database model back in the 70's: https://www.cs.cmu.edu/~pavlo/courses/fall2013/static/papers/whatgoesaround-stonebraker.pdf